File: solutions/java/triangle/1/src/main/java/Triangle.java
class Triangle {
    private double a;
    private double b;
    private double c;
    
    Triangle(double a, double b, double c) throws TriangleException {
        if (a > 0 && b > 0 && c > 0) {
            if (a + b >= c && b + c >= a && a + c >= b) {
                this.a = a;
                this.b = b;
                this.c = c;
            } else {
                throw new TriangleException();
            }
        } else {
            throw new TriangleException();
        }
        }

    boolean isEquilateral() {
        return a == b && b == c;
    }

    boolean isIsosceles() {
        return a == b || b == c || c == a;
    }

    boolean isScalene() {
        return a != b && b!=c && c != a;
    }

}

File: solutions/c/triangle/1/triangle.c
#include "triangle.h"
#include <math.h>
#include <stdbool.h>

bool validate(triangle_t triangle) {
    double a = triangle.a;
    double b = triangle.b;
    double c = triangle.c;
    return (a+b >= c) && (b+c >= a) && (a + c >= b) && !(a == 0 || b == 0 || c==0);
}

bool is_equilateral(triangle_t triangle) {
    if (!validate(triangle)) return false;
    return (triangle.a == triangle.b) && (triangle.b == triangle.c);
}

bool is_isosceles(triangle_t triangle) {
    if (!validate(triangle)) return false;
    return (triangle.a == triangle.b) || (triangle.b == triangle.c) || (triangle.a == triangle.c);
}
bool is_scalene(triangle_t triangle) {
    if (!validate(triangle)) return false;
    return (triangle.a != triangle.b) && (triangle.b != triangle.c) && (triangle.a != triangle.c);
}

